### PR TITLE
[1244] Add support for grid layout on FormDescriptions

### PR DESCRIPTION
**Pull request by AxelRICHARD**
# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [x] Tests
- [ ] Build / releng

## Project management
- [x] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [x] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [x] Have the relevant issues been added to the pull request?
- [x] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [x] Have the relevant issues been added to the same project and milestone as the pull request?
- [x] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [x] Does the title of the commit contributing the ADR start with `[doc]`?
- [x] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [x] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [x] Variables have a proper type
- [x] Functions’ arguments have a proper type
- [x] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [x] `useState<STATE_TYPE>(…)`
- [x] All components have a proper typing for their props
- [x] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [x] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [x] Are data structure classes properly separated from behavioral classes?
- [x] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+  return (
+    <div>
+      <PropertySectionLabel label={widget.label} subscribers={[]} />
+      <Grid
+        className={classes.root}
+        data-testid={widget.label}
+        container
+        onFocus={onFocus}
+        onBlur={onBlur}
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:118]: Are the `onFocus` and `onBlur` pertinent for the grid?

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.types.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+export interface GQLUpdateWidgetFocusInput {
+  id: string;
+  editingContextId: string;
+  representationId: string;
+  widgetId: string;
+  selected: boolean;
+}
+
+export interface GQLUpdateWidgetFocusSuccessPayload extends GQLUpdateWidgetFocusPayload {}
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.types.ts:49]: remove the success payload since it is not used

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx (hunk 1 above)
Comment (AxelRICHARD) [reply frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:118]: No you are right, the WidgetContainerPropertySection has no need to be focusable on the Form representation.

File: backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/WidgetContainerComponentProps.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import org.eclipse.sirius.components.representations.IProps;
+import org.eclipse.sirius.components.representations.VariableManager;
+
+/**
+ * The properties of the widget container component.
+ *
+ * @author arichard
+ */
+public class WidgetContainerComponentProps implements IProps {
+    private VariableManager variableManager;
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/WidgetContainerComponentProps.java:27]: Done

File: backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/WidgetContainerComponent.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import org.eclipse.sirius.components.representations.VariableManager;
+
+/**
+ * The component used to create the widget container widget.
+ *
+ * @author arichard
+ */
+public class WidgetContainerComponent implements IComponent {
+
+    private WidgetContainerComponentProps props;
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/WidgetContainerComponent.java:37]: Done

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.types.ts (hunk 2 above)
Comment (AxelRICHARD) [reply frontend/src/properties/propertysections/WidgetContainerPropertySection.types.ts:49]: Done

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    flexGrow: 1,
+  },
+}));
+
+const getLayout = (layout: WidgetContainerLayout): GridDirection => {
+  if (layout === WidgetContainerLayout.HORIZONTAL) {
+    return 'row';
+  }
+  return 'column';
+};
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:32]: No needed anymore

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  const layout = getLayout(widget.layout);
+
+  return (
+    <div>
+      <PropertySectionLabel label={widget.label} subscribers={[]} />
+      <Grid
+        className={classes.root}
+        data-testid={widget.label}
+        container
+        spacing={2}
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:53]: We don't want to set default values in the frontend. If you really need to set a default value, make it available in the view DSL and set the default there.

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        direction={layout}
+        justifyContent="flex-start"
+      >
+        {widget.children.map((widget) => {
+          return (
+            <Grid item xs key={widget.id}>
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:59]: replace `Grid` by `React.Fragment` (and remove `item` and `xs`)

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+  return (
+    <div>
+      <PropertySectionLabel label={widget.label} subscribers={[]} />
+      <Grid
+        className={classes.root}
+        data-testid={widget.label}
+        container
+        spacing={2}
+        direction={layout}
+        justifyContent="flex-start"
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:55]: same comment here.

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  editingContextId,
+  formId,
+  widget,
+  widgetSubscriptions,
+  setSelection,
+  readOnly,
+}: WidgetContainerPropertySectionProps) => {
+  const classes = useWidgetContainerPropertySectionStyles();
+
+  const layout = getLayout(widget.layout);
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:44]: not needed anymore

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls [hunk 10]
```diff
@@ -213,6 +213,19 @@ type ChartWidget implements Widget {
   chart: Chart!
 }
 
+type WidgetContainer implements Widget {
+  id: ID!
+  diagnostics: [Diagnostic!]!
+  label: String!
+  layout: WidgetContainerLayout!
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls:220]: Done

File: backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls [hunk 11]
```diff
@@ -213,6 +213,19 @@ type ChartWidget implements Widget {
   chart: Chart!
 }
 
+type WidgetContainer implements Widget {
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls:216]: Done

File: backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls [hunk 12]
```diff
@@ -213,6 +213,19 @@ type ChartWidget implements Widget {
...
+  diagnostics: [Diagnostic!]!
+  label: String!
+  layout: WidgetContainerLayout!
+  children: [Widget!]! 
+}
+
+enum WidgetContainerLayout {
+  VERTICAL
+  HORIZONTAL
+}
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-collaborative-forms/src/main/resources/schema/form.graphqls:227]: Done

File: doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+
+=== Frontend
+
+The MUI Grid widget will be used to handle those new widgets.
+
+=== Backend
+
+The View DSL will be modified, introducing the following concepts:
+- new EClass : `WidgetContainerDescription -> WidgetDescription`
+- new EEnum : `WidgetContainerLayout` containing two literals : VERTICAL & HORIZONTAL.
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc:25]: Done

File: doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc [hunk 14]
```diff
@@ -0,0 +1,45 @@
+= ADR-060 - Add support for grid layout on form description
+
+== Context
+
+In the View DSL, it is now possible to create form descriptions, in addition to diagram descriptions.
+A form description contains a set of widgets (textfields, checkboxes, radio buttons...).
+We would like to be able to add grid layout (vertical/horizontal) on those widgets.
+The FormDescriptionEditor representation should also be able to handle those new layouts.
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc:8]: Done

File: doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc [hunk 15]
```diff
@@ -0,0 +1,45 @@
+= ADR-060 - Add support for grid layout on form description
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc:1]: Done

File: doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+== Context
+
+In the View DSL, it is now possible to create form descriptions, in addition to diagram descriptions.
+A form description contains a set of widgets (textfields, checkboxes, radio buttons...).
+We would like to be able to add grid layout (vertical/horizontal) on those widgets.
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc:7]: Done

File: doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+- new EEnum : `WidgetContainerLayout` containing two literals : VERTICAL & HORIZONTAL.
+
+The View DSL converter will be updated to take into account the new widgets.
+
+The new `WidgetContainerDescription` :
+
+```
+	public final class WidgetContainerDescription extends WidgetDescription {
+
+    		private WidgetContainerLayout layout;
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc:34]: Done

File: frontend/src/form/Form.types.ts [hunk 18]
```diff
@@ -176,3 +176,14 @@ export interface ChartWidget extends Widget {
   label: string;
   chart: Chart;
 }
+
+export interface WidgetContainer extends Widget {
+  label: string;
+  layout: WidgetContainerLayout;
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/form/Form.types.ts:182]: Done

File: doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+=== Backend
+
+The View DSL will be modified, introducing the following concepts:
+- new EClass : `WidgetContainerDescription -> WidgetDescription`
+- new EEnum : `WidgetContainerLayout` containing two literals : VERTICAL & HORIZONTAL.
+
+The View DSL converter will be updated to take into account the new widgets.
+
+The new `WidgetContainerDescription` :
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_grid_layout_on_form_description.adoc:29]: Done

File: CHANGELOG.adoc [hunk 20]
```diff
@@ -31,6 +32,7 @@
 - https://github.com/eclipse-sirius/sirius-components/issues/1203[#1203] [charts] Add support for d3 bars data-structure and bar-chart representation
 - https://github.com/eclipse-sirius/sirius-components/issues/1228[#1228] [charts] Add support for bar-chart in view DSL
 - https://github.com/eclipse-sirius/sirius-components/issues/1248[#1248] [charts] Add support for pie-chart in Form representation
+- https://github.com/eclipse-sirius/sirius-components/issues/1244[#1244] [form] Add support for grid layout on FormDescriptions
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply CHANGELOG.adoc:35]: Done

File: CHANGELOG.adoc [hunk 21]
```diff
@@ -5,6 +5,7 @@
 === Architectural decision records
 
 - [ADR-60] Add support for d3 bars data-structure and bar-chart representation
+- [ADR-61] Add support for grid layout on form description
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply CHANGELOG.adoc:8]: Done

File: backend/sirius-components-view/src/main/resources/model/view.ecore [hunk 22]
```diff
@@ -270,6 +270,25 @@
     <eStructuralFeatures xsi:type="ecore:EReference" name="conditionalStyles" upperBound="-1"
         eType="#//ConditionalRadioDescriptionStyle" containment="true"/>
   </eClassifiers>
+  <eClassifiers xsi:type="ecore:EClass" name="BarChartDescription" eSuperTypes="#//WidgetDescription">
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply backend/sirius-components-view/src/main/resources/model/view.ecore:273]: Done

File: frontend/src/form/FormEventFragments.types.ts [hunk 23]
```diff
@@ -234,7 +234,19 @@ export interface GQLLink extends GQLWidget {
...
 }
+
 export interface GQLChartWidget extends GQLWidget {
   label: string;
   chart: GQLChart;
 }
+
+export interface GQLWidgetContainer extends GQLWidget {
+  label: string;
+  layout: GQLWidgetContainerLayout;
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/form/FormEventFragments.types.ts:245]: Done

File: frontend/src/form/FormEventFragments.types.ts [hunk 24]
```diff
@@ -234,7 +234,19 @@ export interface GQLLink extends GQLWidget {
...
+  children: GQLWidget[];
+}
+
+export enum GQLWidgetContainerLayout {
+  VERTICAL = 'VERTICAL',
+  HORIZONTAL = 'HORIZONTAL',
+}
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/form/FormEventFragments.types.ts:252]: Done

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+}));
+
+const getLayout = (layout: WidgetContainerLayout): GridDirection => {
+  if (layout === WidgetContainerLayout.HORIZONTAL) {
+    return 'row';
+  }
+  return 'column';
+};
+
+export const WidgetContainerPropertySection = ({
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:34]: Done

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+  editingContextId,
+  formId,
+  widget,
+  widgetSubscriptions,
+  setSelection,
+  readOnly,
+}: WidgetContainerPropertySectionProps) => {
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:41]: Done

File: frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx (hunk 5 above)
Comment (AxelRICHARD) [reply frontend/src/properties/propertysections/WidgetContainerPropertySection.tsx:32]: Done

File: frontend/src/properties/propertysections/FlexboxContainerPropertySection.tsx [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+  FlexboxContainerPropertySectionStyleProps,
+} from 'properties/propertysections/FlexboxContainerPropertySection.types';
+import { PropertySectionLabel } from 'properties/propertysections/PropertySectionLabel';
+import React from 'react';
+
+const useWidgetContainerPropertySectionStyles = makeStyles<Theme, FlexboxContainerPropertySectionStyleProps>(
+  (theme) => ({
+    children: {
+      display: 'flex',
+      flexWrap: 'wrap',
```
Comment (gcoutable) [inline frontend/src/properties/propertysections/FlexboxContainerPropertySection.tsx:26]: We don't want to set such a value in the frontend. If this value is needed, it should be configurable at least in the flexbox container description in the backend.

**Review by gcoutable — CHANGES_REQUESTED** [review]

File: frontend/src/properties/propertysections/FlexboxContainerPropertySection.tsx (hunk 27 above)
Comment (AxelRICHARD) [reply frontend/src/properties/propertysections/FlexboxContainerPropertySection.tsx]: Done

File: doc/adrs/061_add_support_for_flexbox_container_on_form_description.adoc [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+== Decision
+
+As a first step, the FormDescription will be modified to handle this new widget.
+As a second step, the FormDescriptionEditor will be modified to handle this new widget.
+Flexbox Container should be considered as new widget, containing other widgets.
+No style/conditional style have to be applied on this new widget.
+
+=== Frontend
+
+The MUI Grid widget will be used to handle this new widget (or plain flexbox css if Grid doesn't add any value).
```
[the remark this answers is not in the document]
Comment (AxelRICHARD) [reply doc/adrs/061_add_support_for_flexbox_container_on_form_description.adoc]: Done

**Review by gcoutable — APPROVED** [review]

**Pull request merged**